test: add missing tests to `ndarray/base/unary`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds tests for `ndarray/base/unary` from 0-dimensional up to N-dimensional ndarray.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/16

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
test: add tests upto 3d

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add 4d, 5d, and 6d tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add tests upto nd

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add edge cases

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
chore: cleanup

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.0d.js:
```diff
@@ -0,0 +1,87 @@
+	t.end();
+
+	function scale( z ) {
+		return new Complex128( z.re*10.0, z.im*10.0 );
```
Use `real( z )` and `imag( z )`. We use functional accessors to abstract away direct property access. This ensures that, should we ever change the property names, we only need update the code in one spot in the codebase for everything to work as expected.

kgryte replied:
This applies here and in other files.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.1d.js:
```diff
@@ -0,0 +1,116 @@
+	t.end();
+
+	function scale( z ) {
+		return new Complex128( z.re*10.0, z.im*10.0 );
```
Same comment.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+	t.end();
+
+	function scale( z ) {
+		return new Complex128( z.re*10.0, z.im*10.0 );
```
Same comment here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+
+	unary( [ x, y ], scale );
+
+	expected = new Complex128Array([
```
This test result is unexpected. Why are all the imaginary components zero?

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+	st = shape2strides( sh, ord );
+	o = strides2offset( sh, st );
+
+	x = ndarray( dt, oneTo( numel( sh ), dt ), sh, st, o, ord );
```
Ah, right, I don't think you want to use `oneTo` for complex number arrays. Otherwise, all your imaginary components will be zero and you won't be testing whether the implementation is actually appropriately handling complex number arrays. You could do

```suggestion
	x = ndarray( dt, new Complex128Array( oneTo( numel( sh )*2, 'float64' ) ), sh, st, o, ord );
```

kgryte replied:
This same comment will apply here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+
+	unary( [ x, y ], scale );
+
+	expected = new Complex128Array([
```
This is a good example of where the use of `oneTo` for creating a complex number array is not great, as it makes it harder to determine which `0.0` is expected and which is not.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+	unary( [ x, y ], scale );
+
+	expected = new Complex128Array( x.length*2 );
+	gfill( x.length, new Complex128( 10.0, 0.0 ), expected, st[ 1 ] );
```
When you update L566, you should update this accordingly.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1173 @@
+	unary( [ x, y ], scale );
+
+	expected = new Complex128Array( x.length*2 );
+	gfill( x.length, new Complex128( 10.0, 0.0 ), expected, st[ 1 ] );
```
Same comment.

kgryte reviewed (changes requested):
@headlessNode Left some initial comments. Overall, looking good; however, some clean-up is necessary for how we generate and test complex number arrays.

headlessNode committed:
refactor: apply code review suggestions

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte changes applied, please check.

kgryte commented on lib/node_modules/@stdlib/ndarray/base/unary/test/test.2d.js:
```diff
@@ -0,0 +1,1175 @@
+	unary( [ x, y ], scale );
+
+	expected = new Complex128Array( x.length*2 );
+	gfill( x.length, new Complex128( 10.0, 10.0 ), expected, st[ 0 ] );
```
For future reference, we also have `@stdlib/blas/ext/base/zfill`, which is a package you added. :)

This works. It will just be a bit slower.

kgryte reviewed (approved):
LGTM.
